File: fsharp/word-search/WordSearch.fs
module WordSearch

let search (grid:string list) wordsToSearchFor = 
    let directions = [(-1,-1);(0,-1);(1,-1);(1,0);(1,1);(0,1);(-1,1);(-1,0);]

    let isValid (x,y) = x >= 0 && y >= 0 && x < grid.[0].Length && y < grid.Length

    let rec crawler (x,y) (dx,dy) chars =
        match chars with
        | [c] when grid.[y].[x] = c -> Some (x,y)
        | c::tail when grid.[y].[x] = c ->
            let next = (x + dx, y + dy)
            if isValid next then crawler next (dx,dy) tail
            else None
        | _ -> None

    let crawl (x,y) (word:string) = 
        let chars = word |> Seq.toList
        let found = directions |> List.map (fun dir -> crawler (x,y) dir chars) |> List.choose id
        match found with
        | [(x2,y2)] -> Some (word, Some ((x + 1, y + 1),(x2 + 1, y2 + 1)))
        | _ -> None

    let crawlAll (x,y) = wordsToSearchFor |> List.map (crawl (x,y)) |> List.choose id

    let result = [0..grid.Length-1] |> List.map (fun y -> [0..grid.[0].Length-1] |> List.map (fun x -> crawlAll (x,y)))
    let foundMap = result |> List.concat |> List.concat |> Map.ofList
    wordsToSearchFor |> List.fold (fun map word -> 
        if Map.containsKey word map |> not then Map.add word None map else map) foundMap

File: go/word-search/word_search.go
package wordsearch

import (
	"fmt"
)

func crawl(wordRem string, x, y, dx, dy int, puzzle []string) (*[2]int, bool) {
	if y < 0 || y >= len(puzzle) || x < 0 || x >= len(puzzle[y]) {
		return nil, false
	}
	if puzzle[y][x] != wordRem[0] {
		return nil, false
	}
	if len(wordRem) == 1 {
		return &[2]int{x, y}, true
	}
	return crawl(wordRem[1:], x+dx, y+dy, dx, dy, puzzle)
}

func crawler(word string, x, y int, puzzle []string) (*[2][2]int, bool) {
	if puzzle[y][x] != word[0] {
		return nil, false
	}
	for dx := -1; dx <= 1; dx++ {
		for dy := -1; dy <= 1; dy++ {
			if dx == dy && dx == 0 {
				continue
			}
			res, success := crawl(word[1:], x+dx, y+dy, dx, dy, puzzle)
			if success {
				return &[2][2]int{
					{x, y},
					(*res),
				}, true
			}
		}
	}
	return nil, false
}

func Solve(words []string, puzzle []string) (map[string][2][2]int, error) {
	result := make(map[string][2][2]int)
	for y := 0; y < len(puzzle); y++ {
		for x := 0; x < len(puzzle[y]); x++ {
			for _, w := range words {
				if _, exists := result[w]; exists {
					continue
				}
				res, success := crawler(w, x, y, puzzle)
				if success {
					result[w] = *res
				}
			}
		}
	}
	for _, w := range words {
		if _, exists := result[w]; !exists {
			return result, fmt.Errorf("word '%s' was not found", w)
		}
	}
	return result, nil
}
